In [ ]:
// Import the required libraries
#r "nuget: Plotly.NET.Interactive, 3.0.2"
#r "nuget: FSharp.Stats, 0.4.7"
#r "nuget: FsODE, 0.0.1"

// ... and open the required modules
open FsODE
open Plotly.NET
open Plotly.NET.LayoutObjects
open Plotly.NET.StyleParam
open FSharp.Stats

Installed Packages FSharp.Stats, 0.4.7 FsODE, 0.0.1 Plotly.NET.Interactive, 3.0.2

# Tag 2 - Modellierung von Wachstum

Wiederholung Wachstumsmodellierung

## Exkurs Logarithmus


Wie Sie vielleicht gemerkt haben, wächst die Population in obigem Beispiel zu jedem Zeitpunkt mit der gleichen Rate exponentiell. Die Steigungsänderung hängt nie von *t*, sondern nur von der gerade vorhandenen Zellzahl *N* und der Konstante *r* ab (vgl. 2.2). In den Abbildung wirkt es jedoch, als würden sich die Zellen mit fortschreitender Zeit immer schneller teilen. Um diesen Effekt zu vermeiden sollten Zellzahlen grundsätzlich vor einer Wachstumsanalyse logarithmiert werden. Diese Transformation führt dazu, dass exponentielle Verhaltensweisen visuell korrigiert werden.

Ein Beispiel:

Ein Bakterium teilt sich jede Stunde genau ein mal. Nach der ersten Stunde sind es 2, nach der zweiten 4, nach der dritten 8 usw. Dieses Verhalten lässt sich durch $f(t)=2^t$ wie im Folgenden gezeigt darstellen.

_Aufgabe 2.8: Formulieren Sie die Funktion f(t), die dieses Verhalten widerspiegelt._


In [ ]:
let f t = 
    2.0**t //aufgabe

let xyWerte = 
    [|
        for t in [|0..24|] do  
            let xWert = t
            let yWert = f t
            xWert,yWert
    |]

Chart.Line xyWerte
|> Chart.withXAxisStyle("Time [h]")
|> Chart.withYAxisStyle("Bakterienzahl")
|> Chart.withTitle("Abbildung 4 - Logarithmus I")

<!-- Plotly chart will be drawn inside this DIV -->

Allein durch das Betrachten einer exponentielle Funktion wird nicht klar, was zwischen Stunde 0 und Stunde 15 passiert. Um dafür einen Eindruck zu erhalten werden die y-Werte logarithmiert und erneut visualisiert.

In [ ]:
let werteTransformation (transformation: float -> float) = 
    xyWerte
    |> Seq.map (fun (x,y) -> 
        x,transformation y
        ) 

let log2Transform = werteTransformation log2
let log10Transform = werteTransformation log10
let lnTransform = werteTransformation log

[
Chart.Line (log2Transform,Name="log2")
Chart.Line (log10Transform,Name="log10")
Chart.Line (lnTransform,Name="ln")
]
|> Chart.combine
|> Chart.withXAxisStyle("Time [h]")
|> Chart.withYAxisStyle("logX(Bakterienzahl)")
|> Chart.withTitle("Abbildung 4 - Logarithmus II")


<!-- Plotly chart will be drawn inside this DIV -->

Eine log-transformation, hat auf hohe Werte einen deutlich größeren Einfluss als auf kleine Werte. Die Interpretation des Verlaufs ist nun an jedem beliebigen x-Wert möglich, ohne dass man in den Plot "reinzoomen" muss. Aus exponentiellem Verhalten wird durch log-Transformation eine Gerade. Es wird deutlich, dass egal welche log Transformation angewandt wurde (log zur Basis 2, 10 oder der eulerischen Zahl (natürlicher Logarithmus)) Funktionen zu Linien mit unterschiedlichen Steigung werden.

_Aufgabe 2.9:_

  _a) Wie groß muss der y-Achsenabschnitt sein, um eine Verdopplung der Bakterien zu erhalten? Visuelle Abschätzung möglich._ 

  _b) Um Zellverdopplungen zu analysieren, welchen Logarithmus würden Sie nutzen um die visuelle Analyse so einfach wie möglich zu machen und warum?_ 

  _c) Welcher Logarithmus wird bei pH-Werten genutzt und warum könnte dies als sinnvoll erachtet werden um zum Beispiel pH 2 direkt mit pH 5 vergleichen zu können?_

a)

  - Antwort log2:   

  - Antwort log10:  

  - Antwort ln:     

b)

  - Antwort:

c)

  - Antwort:

Es folgt eine kurze Besprechung. Falls Sie schnell vorangekommen sind, können sie bis zur Besprechung folgende Aufgaben lösen:


### Expertenaufgaben
- Wie groß ist der y-Achsenabschnitt im log space (log2, ln und log10) um von y=100 zu y=10000 zu kommen? Lösen Sie die Aufgabe erst durch schätzen, dann rechnerisch.


___


## Biologisches Beispiel

Von einer Bakterienkultur wurde in regelmäßigen Abständen die Zellzahl bestimmt. Die gemessenen Werte können Sie aus untenstehender Tabelle entnehmen


|Zeitpunkt $(h)$|Zellzahl $(\frac{10^6}{ml})$|
|--|--|
|0.0|17|
|0.5|16.5|
|1.0|11|
|1.5|14|
|2.0|27|
|2.5|40|
|3.0|120|
|3.5|300|
|4.0|450|
|4.5|1,200|
|5.0|2,700|
|5.5|5,000|
|6.0|8,000|
|6.5|11,700|
|7.0|12,000|
|7.5|13,000|
|8.0|12,800|

In der folgenden Zelle wird die Zellzahl als Funktion der Zeit visualisiert.

In [ ]:
let time = [|0. .. 0.5 .. 8.|]

let cellCount =
    [|
    17000000.;16500000.;11000000.;14000000.;27000000.;40000000.;120000000.;
    300000000.;450000000.;1200000000.;2700000000.;5000000000.;
    8000000000.;11700000000.;12000000000.;13000000000.;12800000000.
    |]

let cellCountLog2 = 
    Array.map log2 cellCount

let chartOrig = 
    Chart.Point(time,cellCount)
    |> Chart.withTraceInfo "original data"
    |> Chart.withYAxisStyle "cells*10^6/ml"

let chartLog =
    Chart.Point(time,cellCountLog2)
    |> Chart.withTraceInfo "log count"
    |> Chart.withXAxisStyle "time (h)" 
    |> Chart.withYAxisStyle "log2(cells*10^6/ml)"
    
let growthChart = 
    [chartOrig;chartLog] 
    |> Chart.Grid(2,1)
    |> Chart.withTitle("Abbildung 5 - Beispiel I")

growthChart

<!-- Plotly chart will be drawn inside this DIV -->

Es ist ein deutlicher visueller Unterschied im Funktionsverlauf zu erkennen. Während im nicht-transformierten Diagramm der Zeitpunkt um 6 Stunden noch stark ansteigend (exponentielles Wachstum) erscheint, wird erst im transformierten Diagramm klar, dass die exponentielle Phase von Zeitpunkt 2 - 5 andauert und sich bei Stunde 6 schon am Übergang zum stationären Wachstum befindet. Durch die Abhängigkeit von Steigung und Zellzahl werden die echten Wachstumsphasen in untransformierten Daten kaschiert. Daher sollten Wachstumsdaten in der Regel im log-space angezeigt und analysiert werden. 

Wichtig: Im Folgenden wird immer davon ausgegangen, dass die Daten log transformiert wurden bevor die Kurven modelliert werden!

<center>
<img src="../../images/day2/02_Growth_02.png" width=50% />
</center>

**Abbildung 6: Schema einer klassischen Wachstumskurve**

Wie im Schema gezeigt ist es sinnvoll die Daten zu fitten um bestimmte Parameter wie y0, ymax,  $\mu max$ und $\lambda$ ermitteln zu können. Ein einfaches Modell für exponentielles Wachstum haben wir bereits im ersten Abschnitt kennen gelernt. Würden wir auf die transformierten Daten (Abbildung 5 unten) nun jedoch das einfache Wachstumsmodell anwenden, dann würde es nicht zu einem sigmoidalen Verlauf kommen, sondern die Kurve würde immer weiter exponentiell ansteigen. Es gibt einige [Wachstumsmodelle](http://www.pisces-conservation.com/growthhelp/index.html), die diese Limitationen nicht aufweisen. Gesucht wird eine Erweiterung des einfachen Modells $\frac{dN}{dt} = rN$, die das Wachstum sowohl in der lag-, als auch in der stationären Phase begrenzen kann, also ein Faktor, der abhängig von der maximalen Zellzahl die Wachstumsgeschwindigkeit (growth rate) beschränkt:

$\frac{dN}{dt}=N'=rN(1-\frac{N}{K})$


Gründe biol factors
carrying capacity beispiel aus der vl nennen

wobei $K$ die so genannte carrying capacity, also die Maximalanzahl von Zellen darstellt. Dieses Modell ist das Verhulst-Wachstumsmodell und beschreibt eine logistische Funktion mit unterer und oberer Asymptote.

_Aufgabe 2.10: Erstellen Sie ein Modell der angepassten Differenzialgleichung, einem Wachstumsparameter "r" von 0.3, einer Maximalanzahl "k" von 20 und einer Startzellzahl von 1._





In [ ]:
// parameter
let r = 0.3
let k = 20.

// growth model
let dN_dt : SimpleModel =     
    fun N t ->
        let cellcount = N
        let cellcount' = N * r * (1. - N / k)
        cellcount'

// Initial condition
let t0 = 0
let N0 = 1.

In [ ]:
// Simulate the model function
let sim_dy_dx = 
    modelContext.OdeInt(t0,N0,verhulst)
    |> SolPoints.take 15 
    |> SolPoints.memorize

// Plot the numerical solution
sim_dy_dx
|> SolPoints.toPoints 1
//|> Seq.map (fun (x,y) -> x, log2 y)
|> Chart.Point
|> Chart.withTitle("Abbildung 7 - Simulation verhulst growth model")
|> Chart.withXAxisStyle("Time [h]")
|> Chart.withYAxisStyle("log2(Zellzahl)")

<!-- Plotly chart will be drawn inside this DIV -->

Die lag- und stationäre Phase sind hier deutlich erkennbar. Dieses Modell soll nun an die gegebenen Daten aus dem biologischen Beispiel gefittet werden. Dazu suchen wir die Integration der Differenzialgleichung und versuchen diese Funktion durch Änderung der Parameter nah an die gegebenen logarithmierten Daten zu fitten. Ein Integral lautet:

$$N(t)=lmin + \frac{lmax - lmin}{1+e^{(k-t)/d}}$$

mit

$lmin$ = untere asymptote

$lmax$ = obere asymptote/Zellmaximum

$k$ = x-Wert des Wendepunktes

$d$ = Steilheit

Zur Veranschaulichung wird die Verhulst-Funktion mit 4 verschiedene Parametersets visualisiert.

In [ ]:
let verhulstFunktion lmin lmax (k:float) d t=
    lmin + (lmax - lmin)/(1.+ exp((k-t)/d))

let plotVerhulst lmin lmax  k d =
    let data=
        [0..40]
        |> Seq.map (fun x -> x,verhulstFunktion lmin lmax k d x)
    let chart = 
        Chart.Line(data,Name=sprintf "min:%.0f max:%.0f k:%.0f d:%.0f" lmin lmax k d)
    chart

[
plotVerhulst 0. 10. 4. 1.
plotVerhulst 0. 10. 20. 1.
plotVerhulst 0. 10. 20. 3.
plotVerhulst 5. 20. 10. 3.
]
|> Chart.combine
|> Chart.withLegendStyle(X=0.6,Y=0.01)
|> Chart.withTitle("Abbildung 8 - Verschiedene logistische Funktionen")
|> Chart.withXAxisStyle("Time [h]")
|> Chart.withYAxisStyle("log2(Zellzahl)")

<!-- Plotly chart will be drawn inside this DIV -->

Ein sogenannter Solver benötigt die Daten und das Modell, welches auf diese Daten appliziert werden soll. Der Solver mit dem Namen Levenberg-Marquardt versucht die Parameter des Modells so lange zu optimieren, bis der Abstand zwischen dem Modell und den echten Daten minimal geworden ist. Anschließend kann man die optimierten Parameter abfragen und mit diesen weiter arbeiten. Mehr Details finden Sie in *Vorlesung 13 - Principles of model fitting*.
 

In [ ]:
open System
open FSharp.Stats
open FSharp.Stats.Fitting.NonLinearRegression
open FSharp.Stats.Fitting.LinearRegression

let getVerhulstParams timePoints logdata min max k d = 
    LevenbergMarquardt.estimatedParams
        Table.GrowthModels.verhulst4Param
        (createSolverOption 0.001 0.001 10000 [|max;k;d;min|])
        0.1
        10.
        timePoints
        logdata

let verhulstParams =
    //getVerhulstParams time cellCountLog2 23. 35. 3.5 1.
    getVerhulstParams time cellCountLog2 23.391 33.598 4. 1.

printfn "Durch cleveres Ausprobieren identifiziert der Solver folgende Parameter:
lmin = %.3f 
lmax = %.3f
k = %.3f
d = %.3f" verhulstParams.[3] verhulstParams.[0] verhulstParams.[1] verhulstParams.[2]



Für die gegebenden Daten sind die optimalen Parameter des 4-Parameter Verhulstmodells:
lmin = 23.578 
lmax = 33.987
k = 3.997
d = 0.956


Der folgende Codeblock dient nur der Visualiserung der Daten mit gefittetem Modell.

In [ ]:
let fittedValuesVerhulst =
    /// The parameter were determined locally for saving time during build processes
    let f =  Table.GrowthModels.verhulst4Param.GetFunctionValue verhulstParams
    [time.[0] .. 0.1 .. Seq.last time]
    |> Seq.map (fun x -> 
        x,f x
        ) 
    |> Chart.Line

let fittedChartVerhulst = 
    fittedValuesVerhulst 
    |> Chart.withXAxisStyle "time (h)" 
    |> Chart.withYAxisStyle "log2(cells/ml)"
    |> Chart.withTraceInfo "verhulst"

let fittedChartVerhulstS = 
    [
        Chart.Point(time,cellCountLog2)
        |> Chart.withTraceInfo "log2 count"
        fittedChartVerhulst
    ]
    |> Chart.combine
    |> Chart.withXAxisStyle "time (h)" 
    |> Chart.withYAxisStyle "log2(cells/ml)"

fittedChartVerhulstS

<!-- Plotly chart will be drawn inside this DIV -->

Es folgt eine kurze Besprechung. Falls Sie schnell vorangekommen sind, können sie bis zur Besprechung folgende Aufgaben lösen:


### Expertenaufgaben
- Teilen Sie die Zeitpunkte in lag, log und stationäre Phase ein. (als echte fragen)
- Wie sind im gezeigten Fall die Werte der unteren und oberen Asymptote? Geben Sie die Werte als untransformierte Zellzahl an.

___

## Verdopplungszeit-Bestimmung

Wie bereits herausgefunden, ist eine Zellverdopplung im log2 space gleichzusetzen mit einem y-Achsenabschnitt von 1. Um nun die Generations/Verdopplungs-Zeit zu bestimmen muss die Zeit bestimmt werden in dem sich die log(Zellzahl) um 1 erhöht. Dies kann durch die maximale Steigung ($\mu max$) des gefitteten Modells erreicht werden. Diese entspricht dem schnellstmöglichen Wachstum und ist naturgemäß am Wendepunkt (inflection point) der sigmoidalen Kurve. Diese Steigung wird auch als "maximal growth rate" oder "specific growth rate" bezeichnet und wird gerne genutzt um Unterschiede zwischen Experimenten zu quantifizieren. 

Aber Achtung:

> Two definitions for this parameter [$\mu max$ or maximum growth rate] are in current use. One is based on the inflection of the slope of the growth curve in the exponential phase (Zwietering et al., 1990) whilst the other is taken to be the growth rate at infinite dilution (Baranyi et al., 1993). - _Perni et al., 2004_


_Aufgabe 2.11: Schätzen Sie ab, wo sich der Wendepunkt - also der Punkt maximaler Steigung - befindet._

Antwort:

Da im Verhulstmodell ein Parameter schon optimiert wurde, der genau den x-Wert des Wendepunktes definiert, muss man nur noch diesen Wert (k = 3.525) in die Ableitung einsetzen um die maximale Steigung zu erhalten

$$N(t)=lmin + \frac{lmax - lmin}{1+e^{(k-t)/d}}$$


$$N'(t)=\frac{(lmax-lmin)e^{(k-t)/d}}{d(e^{(k-t)/d}+1.)^2}$$

optimierte Parameter aus Model-Fitting:

$lmin = 23.578$

$lmax = 33.987$

$k = 3.997$

$d = 0.956$

$$N(3.997) = 28.782$$

$$N'(3.997) = \frac{lmax - lmin}{4d} = 2.72$$


The maximal slope was determined to be 2.72. 

In [ ]:
let generationtimeVerhulst (verhulstParameters:vector) =
    let lmax = verhulstParameters.[0] //lmax = 33.696
    let k    = verhulstParameters.[1] //k = 3.525
    let d    = verhulstParameters.[2] //d = 1.073
    let lmin = verhulstParameters.[3] //lmin = 22.706
    let gradientFunctionVerhulst t =
        let numerator = (lmax-lmin)*Math.Exp((k-t)/d)
        let denominator = d*(Math.Exp((k-t)/d)+1.)**2.
        numerator / denominator
    let maximalSlope =
        gradientFunctionVerhulst k
    log2(2.) / maximalSlope

let generationVerhulst = 
    sprintf "The generation time (Verhulst) is: %.1f min" (generationtimeVerhulst verhulstParams * 60.)

printfn "%s" generationVerhulst

The generation time (Verhulst) is: 22.0 min


Die bestimmte Verdopplungszeit von 22 Minuten kann nun genutzt werden um sie gegen Verdopplungszeiten anderer Experimente zu vergleichen.

### Anleitung Generationszeitbestimmung
1. Plotten Sie die log2 transformierten Daten als xy Diagramm.
2. Wenn ein Wachstumsmodell und Solver vorhanden ist, fitten Sie das Modell an die Daten und isolieren sie $\mu max$.
3. Wenn kein Wachstumsmodell vorhanden ist, schätzen Sie den linearen Bereich der Wachstumskurve.
    - Erstellen Sie einen Ausgleichsgerade von diesem Bereich und isolieren Sie die Steigung $\mu max$.
4. Um die Generationszeit zu bestimmen rechnen Sie $\frac{log_2 2.}{\mu max} = \frac{1}{\mu max}$.


### Take home messages
- Die logistische Wachstumsfunktion $N(t)=lmin + \frac{lmax - lmin}{1+e^{(k-t)/d}}$ hat niemand einfach entwickelt, sondern es ist ein Integral eines leicht nachzuvollziehenden Wachstumsmodells $\frac{dN}{dt}=rN(1-\frac{N}{k})$
- Wenn Wachstum untersucht werden soll, sollte man sich den zeitlichen Verlauf (fast) immer mit log-transformierten Zellkonzentrationen anschauen.
- Bei log2 transformierten Daten ist ein y-Achsenabschnitt von 1 gleichzusetzen mit einer Verdopplung der Zellzahl, bei einer log10 Transformation ist es eine Verzehnfachung.
- Die Verdopplungszeit hängt ausschließlich von der Steigung der log-phase ($\mu max$ oder _maximal growth rate_) ab. Sie wird berechnet mit $\frac{log_x 2}{\mu max}$ wobei x die gewählte Basis ist, mit der die Zellzahlen transformiert wurden.
- Die Steigung kann entweder durch manuelle Selektion der log phase und anschließendes Fitten einer Gerade geschehen, oder durch fitten eines Wachstumsmodells ([Gompertz/Richards/Verhulst](https://fslab.org/FSharp.Stats/GrowthCurve.html)).



### References
- Maier R., Pepper I., Chapter 3 - Bacterial Growth, Environmental Microbiology (Third Edition), 2015, https://doi.org/10.1016/B978-0-12-394626-3.00003-X
- Perni S., Andrew P., Shama G., Estimating the maximum growth rate from microbial growth curves: definition is everything, Food Microbiology, 2005, https://doi.org/10.1016/j.fm.2004.11.014
- Venn B., Schneider K., Weil L., Zimmer D. and Mühlhaus T., fslaborg/FSharp.Stats: Release 0.4.7, 2022, Growth curve documentation: https://fslab.org/FSharp.Stats/GrowthCurve.html
- Growth II, version 2.3.6.71, PISCES Conservation Ltd United Kingdom



### Expertenaufgaben
- Wie ist die Verdopplungszeit? Die Zeipunkte sind als Tage angeben.
  - a) $\mu max = 2.5$; Basis der log-Transformation: 2
  - b) $\mu max = 1.0$; Basis der log-Transformation: e
  - c) $\mu max = 0.7$; Basis der log-Transformation: 10
  - d) $\mu max = 10$; Basis der log-Transformation: e
  - e) $\mu max = 10$; Basis der log-Transformation: 2
  - f) $\mu max = 10$; Basis der log-Transformation: 10


